esp32/modfreedomgfx.c

```diff
@@ -0,0 +1,273 @@
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(gfx_deinit_obj, gfx_deinit);
+
+STATIC mp_obj_t gfx_poll(void) {
+  uint32_t evt = freedomgfxPoll();
+  gfx_input_poll(evt);
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(gfx_poll_obj, gfx_poll);
+
+STATIC mp_obj_t gfx_line(mp_uint_t n_args, const mp_obj_t *args) {
+	int x0 = mp_obj_get_int(args[0]);
+	int y0 = mp_obj_get_int(args[1]);
+	int x1 = mp_obj_get_int(args[2]);
+	int y1 = mp_obj_get_int(args[3]);
+	int col = mp_obj_get_int(args[4]);
+
+	PXb(x0,y0);
+
+	// algorithm: https://de.wikipedia.org/wiki/Bresenham-Algorithmus
+	int dx =  ABS(x1-x0);
+	int sx = x0<x1 ? 1 : -1;
+	int dy = -ABS(y1-y0);
+	int sy = y0<y1 ? 1 : -1;
+	int err = dx+dy;
+	int err2;
+
+	while(1){
+		if(col == 0) PXb(x0,y0); else PXw(x0,y0);
+		if ((x0 == x1) && (y0 == y1))
+			break;
+		err2 = 2*err;
+		if (err2 > dy)
+		{
+			err += dy;
+			x0 += sx;
+		}
+		if (err2 < dx)
+		{
+			err += dx;
+			y0 += sy;
+		}
+	}
+
+	return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_line_obj, 5, 5, gfx_line);
+
+STATIC mp_obj_t gfx_area(mp_uint_t n_args, const mp_obj_t *args) {
+  int x0 = mp_obj_get_int(args[0]);
+  int y0 = mp_obj_get_int(args[1]);
+  int a = mp_obj_get_int(args[2]);
+  int b = mp_obj_get_int(args[3]);
+  int col = mp_obj_get_int(args[4]);
+
+  for(int i = 0; i < a; i++)
+  {
+	  for(int j = 0; j < b; j++)
+	  {
+		  if(col == 0)
+			  PXb(x0+i,y0+j);
+		  else
+			  PXw(x0+i,y0+j);
+	  }
+  }
+
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_area_obj, 5, 5, gfx_area);
+
+
+static uint8_t tm12x6_font[] = {
+#include "../share/fonts/Lat2-Terminus12x6.inc"
+};
+
+STATIC mp_obj_t gfx_string(mp_uint_t n_args, const mp_obj_t *args) {
+  mp_uint_t len;
+  const unsigned char *data = (const unsigned char*) mp_obj_str_get_data(args[2], &len);
+  int x0 = mp_obj_get_int(args[0]);
+  int y0 = mp_obj_get_int(args[1]);
+  int col = mp_obj_get_int(args[4]);
+
+  uint8_t* font = tm12x6_font;
+  int clen = ((int*)font)[5];
+  int cheight = ((int*)font)[6];
+  int cwidth = ((int*)font)[7];
+
+  int xoffs = 0;
+  while(*data != 0)
+  {
+	  for(int i = 0; i < cheight; i++)
+	  {
+		  for(int j = 0; j < cwidth; j++) // TODO: only works for single byte rows
+		  {
+			  if( font[4*8+(int)(*data)*clen+i] & (1<<(7-j)))
+			  {
+				  if(col == 0)
+					  PXb(x0+j+xoffs,y0+i);
+				  else
+					  PXw(x0+j+xoffs,y0+i);
+			  }
+		  }
+	  }
+	  data++;
+	  xoffs += cwidth;
+
+	  if(*data == '\n')
+	  {
+		  xoffs = 0;
+		  y0 += cheight;
+		  data++;
+	  }
+  }
+
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_string_obj, 5, 5, gfx_string);
+
+
+STATIC mp_obj_t gfx_flush(mp_uint_t n_args, const mp_obj_t *args) {
+	freedomgfxDraw();
+	return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(gfx_flush_obj, 0, 1, gfx_flush);
+
+
+// callback system
+
+STATIC uint8_t button_init_done = 0;
+STATIC mp_obj_t button_callbacks[1+BADGE_BUTTONS];
```

Comment:
(minor)
On some embedded platforms, the .bss section is not initialized. This doesn't matter on the badge platform, but I would still suggest adding '= { NULL }' to force the .data section on these platforms. (by default, gcc places zero-filled variables in the .bss section, so this would not change anything on the badge platform)